### PR TITLE
Update to Dogma 0.2.0

### PR DESCRIPTION
**Pull request by jmalloc**
This change updates Dogma version to 0.2.0, which means the memory engine needs to  implement the `Key()` and `Time()` methods to `ProjectionEventScope`.

In order to implement `Time()`, I had to add a `Time` field to `envelope.Envelope`, which necessitated some fairly wide-reaching changes through the controller implementations.

### CODE REVIEW

File: engine/envelope/envelope.go [hunk 1]
```diff
@@ -78,17 +85,19 @@ func (e *Envelope) NewTimeout(
 	id string,
 	m dogma.Message,
 	t time.Time,
+	s time.Time,
```
Comment (koden-km) [inline engine/envelope/envelope.go]: What is `t` and `s`? Due to there being two time values, should the docs describe these a bit?

Comment (jmalloc) [reply engine/envelope/envelope.go]: Yeah, you're right.

I think I'll rename the struct fields too:

- `Time` -> `CreatedAt`
- `TimeoutTime` -> `ScheduledFor`

Or something like that.

Comment (koden-km) [reply engine/envelope/envelope.go]: That sounds much clearer.

**Comment by jmalloc** [comment]
Ok, this PR crept in scope a bit - I took it as an opportunity to improve the interface of `Envelope` as it's grown a bit.

- Added new `CreatedAt` time
- Renamed `TimeoutTime` to `ScheduledFor` and also change it from a `*time.Time` to just `time.Time`
- Ditched the `envelope.New()` function in favour of explicit `NewCommand()` and `NewEvent()` functions (to avoid passing roles around in the API).

**Pull request merged**